hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+
+The first kind is when a 'real' failure is going to happen in the future. This 'real' failure is the act of decommissioning. We denote this as "decommission" throughout this paper. The response that the user wants is SCM/Ozone to make replicas to deal with the planned failure.
+
+The second kind is when the failure is 'transient.' The user knows that this failure is temporary and cluster in most cases can safely ignore this issue. However, if the transient failures are going to cause a failure of availability; then the user would like the Ozone to take appropriate actions to address it.  An example of this case, is if the user put 3 data nodes into maintenance mode and switched them off.
+
+The transient failure can violate the availability guarantees of Ozone; Since the user is telling us not to take corrective actions. Many times, the user does not understand the impact on availability while asking Ozone to ignore the failure.
+
+So this paper proposes the following definitions for Decommission and Maintenance of data nodes.
+
+__Decommission__ of a data node is deemed to be complete when SCM/Ozone completes the replica of all containers on decommissioned data node to other data nodes.That is, the expected count matches the healthy count of containers in the cluster.
+
+__Maintenance mode__ of a data node is complete if Ozone can guarantee at least one copy of every container is available in other healthy data nodes.
+
+## Examples 
+
+Here are some illustrative examples:
+
+1.  Let us say we have a container, which has only one copy and resides on Machine A. If the user wants to put machine A into maintenance mode; Ozone will make a replica before entering the maintenance mode.
+
+2. Suppose a container has two copies, and the user wants to put Machine A to maintenance mode. In this case; the Ozone understands that availability of the container is not affected and hence can decide to forgo replication.
+
+3. Suppose a container has two copies, and the user wants to put Machine A into maintenance mode. However, the user wants to put the machine into maintenance mode for one month. As the period of maintenance mode increases, the probability of data loss increases; hence, Ozone might choose to make a replica of the container even if we are entering maintenance mode.
+
+4. The semantics of decommissioning means that as long as we can find copies of containers in other machines, we can technically get away with calling decommission complete. Hence this clarification node; in the ordinary course of action; each decommission will create a replication flow for each container we have; however, it is possible to complete a decommission of a data node, even if we get a failure of the  data node being decommissioned. As long as we can find the other datanodes to replicate from and get the number of replicas needed backup to expected count we are good.
+
```

Comment:
> it is possible to complete a decommission of a data node, even if we get a failure of the  data node being decommissioned
This sounds like a state transition from DECOMMISSIONING -> DEAD, similar to IN_SERVICE -> DEAD. Its not really true that decommissioning completes if the node fails. For example if the node came back online after the failure, and replication of the under-replicated container had not yet completed, then the node would go back to DECOMMISSIONING until all the containers it hosts have reach 3 live replicas.

Reply:
@anuengineer can confirm it, but I think this paragraph is about the minimum required replicas. Node can be decommissioned even if the expected number of replicas are not yet available if the minimum required replicas can guarantee that the replication can be finished without the decommissioning node.

Reply:
It would probably make the decommissioning logic cleaner if we just let the "handle dead node" logic take over, and remove the node from any decommission manager should it go dead while it is being tracked. Otherwise you can have the scenario where decommissioning is completed, but the cluster is still not healthy as it has under-replicated containers. This is somewhat of an edge case, but it can easily happen so we do need to consider how to take care of it.